### ISSUE TITLE
Remove the implicit TLS-passed handles and go back to explicitly passing handles as arguments

### ISSUE THREAD

**Issue by khuey** [tokio-rs/tokio#344]
(I'm happy to write this up in the form of a PR to the tokio-rfcs repo if preferred)

## tl;dr
- Passing handles implicitly through TLS has severe downsides.
- Some use cases require explicit handles.
- There's no way to transition from implicit to explicit handles.
- Libraries have to make decisions on implicit vs. explicit for their downstream consumers, or present terrible APIs.
- The implicit handles greatly enlarge the "reasoning footprint"
- API consistency is good.
- The ergonomics of explicit state are manageable with traits.
- tokio should do that instead and drop the current implicit state model.

## Passing handles implicitly through TLS has severe downsides

The futures context RFC[0] made compelling arguments here. I'd note one thing particularly pertinent to tokio: with the implicitly-passed state, it's impossible to determine which of the different tokio components (a reactor, an executor, or timers) are needed to run a given function from its signature. Similarly, it's impossible to determine which components a library requires without reading its documentation. The problem of not having an environment correctly set up to call a function is even worse for tokio than it was for futures because it consists of multiple separate components that can be mixed and matched.

## Some use cases require explicit handles

To give one example: it is not possible for a library to start its own thread and send events/messages/etc back to the tokio event loop via `tokio::spawn`.  This case requires explicit handles.

## There's no way to transition from implicit to explicit handles

There are no getters that allow one to grab the current executor, or reactor, or timer handle, and pass them around in a form which is known to be usable on a different thread or even at all points on the same thread.

## Libraries have to choose

Even ignoring the use cases that cannot be done with implicit handles, because there is no way to transition from implicit to explicit handles, a library that chooses explicit handles forces all of its consumers to use explicit handles as well. Because `tokio::run`, now the recommended way to start tokio, is entirely implicit, I expect that few libraries will use explicit handles unless they must.

If you did want to make your library agnostic as to where the state is, it would have to take `Option<Handle>` everywhere. And instead of handling one case for explicit argument passing (the argument is present), and two for implicit (the handle is in the TLS slot or it is not), it would have to handle all three.

## The implicit handles great enlarge the "reasoning footprint"

Last year aturon suggested a framework[1] for analyzing "implicit vs explicit" tradeoffs in terms of the "reasoning footprint", measured on the dimensions of:

- **Applicability.** *Where are you allowed to elide implied information? Is there any heads-up that this might be happening?* The implicit handle arguments can be elided anywhere you call into code that uses tokio, and there's no easy way to know which code uses tokio. So this is applicable to a potentially large, unknown set of code.

- **Power.** *What influence does the elided information have? Can it radically change program behavior or its types?* The implicit handle arguments determine whether features work at all, potentially which thread work gets dispatched to, and whether the program panics (if they are missing). So they have large effects on program behavior.

- **Context-dependence.** *How much of do you have to know about the rest of the code to know what is being implied, i.e. how elided details will be filled in? Is there always a clear place to look?* To fully understand what's going on you have to look both down and up all possible call chains to see what handles might be required and provided, respectively. This is extremely context-dependent.

aturon suggested that "if a feature is large in one of the dimensions, it’s best to strongly limit it in the other two." This method of passing arguments is large in all three dimensions.

## API consistency is good

futures and tokio are closely related parts of the async rust ecosystem. It feels extremely weird to have them pulling in opposite directions on fundamental design decisions like this.

## The ergonomics of explicit state are manageable with traits.

Passing handles explicitly doesn't have to mean passing several separate arguments to every function that uses tokio. We could declare traits to represent the presence of the various handles, and then write functions that bind over the corresponding traits they need. For example:

```rust
trait HasReactor {
  fn reactor(&self) -> &reactor::Handle;
}

trait HasExecutor {
  fn executor(&self) -> &executor::Executor;
}

trait HasTimer {
  fn timer(&self) -> &timer::Handle;
}

impl HasReactor for (reactor::Handle, executor::Executor) {
  ...
}
impl HasExecutor for (reactor::Handle, executor::Executor) {
  ...
}

fn do_io<H>(handles: H, arguments ...) -> impl Future<...>
  where H: HasReactor + HasExecutor
{
  ...
}

fn do_io_with_timeout<H>(handles: H, arguments ...) -> impl Future<...>
  where H: HasReactor + HasExecutor + HasTimer
{
  ...
}
```

Then I can call `do_io` but not `do_io_with_timeout` with a single set of handles. In the common case, `Runtime` could produce a cloneable `Handles` object that implements these traits and can be passed into functions. Consumers with more complex needs could implement these traits on whatever object is convenient for them. The type system will verify at compile time that I have provided the proper set of handles for the function in question.

## Proposal

- Introduce the `HasReactor`/`HasExecutor`/`HasTimer` traits described above.
- Add a `Runtime::handles(&self) -> Handles` or similar, where `&Handles: HasReactor + HasExecutor + HasTimer + Clone`. This would combine `Runtime::reactor` and `Runtime::executor` (and `Runtime::timer`, which doesn't actually exist) into a single return value.
- Change `tokio::run` to be
```rustc
fn tokio::run<C, F>(c: C)
    where C: FnOnce(&Runtime) -> F,
          F: IntoFuture,
          F::Future: Future<Item = (), Error = ()> + Send + 'static {
    let mut runtime = Runtime::new().unwrap();
    let f = c(&runtime).into_future();
    runtime.spawn(f);
    runtime.shutdown_on_idle().wait().unwrap();
}
```
- Remove the implicit state: i.e. make `tokio::spawn` take a `H: HasExecutor` argument, make `Registration::register` take a `H: HasReactor` argument, etc.

Then a consumer would look like:
```rustc
use tokio::net::TcpListener;

fn process<H: HasTimer>(h: H, socket: TcpStream) -> impl Future<Item = (), Error = ()> {
    // Do work on the socket or timeout ...
}

tokio::run(|runtime| {
    let h = runtime.handles();
    let listener = TcpListener::bind(&h, &addr).unwrap();

    let server = listener.incoming()
        .map_err(|e| println!("error = {:?}", e))
        .for_each(move |socket| {
            tokio::spawn(&h, process(&h, socket))
        });

    server
})
```

[0] https://github.com/rust-lang-nursery/futures-rfcs/blob/master/task-context.md
[1] https://blog.rust-lang.org/2017/03/02/lang-ergonomics.html

**Comment by carllerche** [comment]
Thanks for the thoughtful post. The proposed changes are breaking and as such cannot be considered until a breaking release. As I have previously mentioned, before the next major breaking release, I will post an RFC which would be a good time for everyone to get involved in the discussion.

I will also post some thoughts now.

### Regarding: Passing handles implicitly through TLS has severe downsides

This point references the RFC to futures. However, the `futures` library and Tokio have significantly different scopes and the explicit task context has different implications than explicit handles do.

First, the `futures` crate is striving to be a fully general purpose abstraction for all asynchronous contexts, including `no_std` cases. Being able to support cases that did not have thread-locals was a significant factor.

In the futures case, the task context is already held by the executor. The proposal was to just pass it in as `&mut self` instead of via TLS. The only cost of doing so was ergonomic and had no real runtime penalty.

Tokio, on the other hand, is heavily `std` focused and uses threads and other std features liberally.

For context, `tokio-core` used explicit `Handle` values everywhere. There was originally a discussion [here](https://github.com/tokio-rs/tokio-core/pull/84). An insightful comment is that requiring an explicit handle argument means that it has to be stored by *every* task instead of once per thread. This can result in significant memory bloat. See [here](https://github.com/tokio-rs/tokio-core/pull/84#issuecomment-270265815).

So, switching from explicit `Handle` values to implicit via TLS comes w/ significant memory benefits.

> Similarly, it's impossible to determine which components a library requires without reading its documentation.

This is a valid concern, but I believe it can be handled via convention instead.

Tokio libraries should, by default, assume that the `Runtime` is present and only implicitly use capabilities provided by the `Runtime`. There should also be a “configuration” path where all dependencies are explicitly passed in.

You will note that Tokio’s own types follow this pattern. For example, `TcpStream::connect` assumes the `Runtime`, but `TcpStream::from_std`, which is the way you configure a TCP socket, takes `&Handle`.

### Regarding: Some use cases require explicit handles

Tokio fully supports using handles when needed.

### Regarding: There's no way to transition from implicit to explicit handles

Yes there is: `Handle::default()`.

### Regarding: Libraries have to choose

I addressed that above w/ “assume the runtime by default and provide a configuration path that is explicit”.

You also do not need any `Option<Handle>`, you can just take a normal `Handle` and use `Handle::default()` to get the default behavior.

### Regarding: The implicit handles great enlarge the "reasoning footprint"

Again, the implicitness can be managed by convention and convention.

* `Applicability`: One should only elide the implied information when using `Runtime`. The heads up is that one is using `Runtime`.

* `Power`: Working vs. not working is not “radical”. It is a boolean. It either works or it does not.

* `Context-dependence`: You have to know if you are within the context of an application that is built around `Runtime` or not.

If you are are not building your app around `Runtime`, you should pass in the dependencies explicitly.

Either way, aturon’s suggestion is just that, a suggestion.

### Regarding: API consistency is good

API consistency is not good only for API consistency sake. As described above, Tokio and `futures` have different goals, as such, things are going to be different.

Also, there are still open questions regarding how well “TLS free” is going to work out for `futures` and there is yet to be significant vetting. See [here](https://github.com/rust-lang-nursery/futures-rs/issues/830).

### Regarding: The ergonomics of explicit state are manageable with traits.

I would not consider the proposal ergonomic compared to the existing API.

Also, adding generics to anything is going to cause trouble w.r.t. trait objects, etc…

### Some APIs **require** thread-locals

The [`blocking`](https://docs.rs/tokio-threadpool/0.1.3/tokio_threadpool/fn.blocking.html) annotation basically requires thread locals. The only way I could think of implementing it without thread locals would be forking all the `Future` traits to pass in additional, custom, context. This seems like a non-starter to me.

**Comment by khuey** [comment]
Thanks for the responses. I look forward to the RFC, and if you want to postpone this until then that's fine by me. A couple quick comments:

- I haven't thought deeply about everything you've written but some of these problems, particularly the million connection case, do seem difficult to do efficiently without some global state stored off to the side.
- `Handle::default()` is not really a good substitute for explicit handles because it can produce a `Handle` that doesn't actually work when you try to use it.
- Why can't `blocking`s closure just capture whatever state it needs that would otherwise be thread local?

**Comment by carllerche** [comment]
> Why can't blockings closure just capture whatever state it needs that would otherwise be thread local?

Because this state is unique to each call to `Future::poll` and must be provided by the executor. The only way to make it work without TLS would be to change `Future` and all other traits:

```rust
trait Future<T: Task> {
    type Item;
    type Error;

    fn poll(&mut self, cx: &mut T) -> Poll<Self::Item, Self::Error>;
}
```

Then, `blocking` would only impl `Future` where `T = TokioTask` or something like that.

Given that `futures` is attempting to stabilize and this would be a significant change, I doubt it could happen.
